simulators/ethereum/engine/clmock.go

```diff
@@ -0,0 +1,365 @@
+
+	The test cases can request a lock to "wake up" at a specific point during
+	the PoS block production procedure and, when the CL mocker has reached it,
+	the lock is released to let only one test case do its testing.
+	*/
+	PayloadBuildMutex                SetResetLock
+	NewGetPayloadMutex               SetResetLock
+	NewExecutePayloadMutex           SetResetLock
+	NewHeadBlockForkchoiceMutex      SetResetLock
+	NewSafeBlockForkchoiceMutex      SetResetLock
+	NewFinalizedBlockForkchoiceMutex SetResetLock
+}
+
+func NewCLMocker() *CLMocker {
+	// Init random seed for different purposes
+	rand.Seed(time.Now().Unix())
+
+	// Create the new CL mocker
+	newCLMocker := &CLMocker{
+		EngineClients:               make([]*EngineClient, 0),
+		RandomHistory:               map[uint64]common.Hash{},
+		ExecutedPayloadHistory:      map[uint64]catalyst.ExecutableDataV1{},
+		LatestFinalizedHeader:       nil,
+		PoSBlockProductionActivated: false,
+		BlockProductionMustStop:     false,
+		FirstPoSBlockNumber:         nil,
+		LatestFinalizedNumber:       nil,
+		TTDReached:                  false,
+		NextFeeRecipient:            common.Address{},
+		LatestForkchoice: catalyst.ForkchoiceStateV1{
+			HeadBlockHash:      common.Hash{},
+			SafeBlockHash:      common.Hash{},
+			FinalizedBlockHash: common.Hash{},
+		},
+	}
+
+	// Lock the mutexes. To be unlocked on next PoS block production.
+	newCLMocker.PayloadBuildMutex.LockReset()
+	newCLMocker.NewGetPayloadMutex.LockReset()
+	newCLMocker.NewExecutePayloadMutex.LockReset()
+	newCLMocker.NewHeadBlockForkchoiceMutex.LockReset()
+	newCLMocker.NewSafeBlockForkchoiceMutex.LockReset()
+	newCLMocker.NewFinalizedBlockForkchoiceMutex.LockReset()
+
+	// Start timer to check when the TTD has been reached
+	time.AfterFunc(tTDCheckPeriod, newCLMocker.checkTTD)
+
+	return newCLMocker
+}
+
+// Add a Client to be kept in sync with the latest payloads
+func (cl *CLMocker) AddEngineClient(newEngineClient *EngineClient) {
+	cl.EngineClients = append(cl.EngineClients, newEngineClient)
+}
+
+// Remove a Client to stop sending latest payloads
+func (cl *CLMocker) RemoveEngineClient(removeEngineClient *EngineClient) {
+	i := -1
+	for j := 0; j < len(cl.EngineClients); j++ {
+		if cl.EngineClients[j] == removeEngineClient {
+			i = j
+			break
+		}
+	}
+	if i >= 0 {
+		cl.EngineClients[i] = cl.EngineClients[len(cl.EngineClients)-1]
+		cl.EngineClients = cl.EngineClients[:len(cl.EngineClients)-1]
+	}
+}
+
+// Check whether we have reached TTD and then enable PoS block production.
+// This function must NOT be executed after we have reached TTD.
+func (cl *CLMocker) checkTTD() {
+	if len(cl.EngineClients) == 0 {
+		// We have no clients running yet, we have not reached TTD
+		time.AfterFunc(tTDCheckPeriod, cl.checkTTD)
+		return
+	}
+
+	// Pick a random client to get the total difficulty of its head
+	ec := cl.EngineClients[rand.Intn(len(cl.EngineClients))]
+
+	lastBlockNumber, err := ec.Eth.BlockNumber(ec.Ctx())
```

Comment:
Is there a reason to calculate this iteratively? You should be able to get a block with `HeaderByNumber` and use the `latest` tag to get the head block, which includes the `totalDifficulty` in the response.

Reply:
For some reason, totalDifficulty is not available in the struct returned by HeaderByNumber:
https://github.com/ethereum/go-ethereum/blob/1884f37f2c772d6f1d567432e76cbd7b90d442a9/core/types/block.go#L69

Neither does the Block struct, so instead of directly calling the RPC and then parsing the response, I calculated the difficulty manually.

Reply:
Ah I see, didn't realize there was no type `RPCHeader`. I think you could do something like this though?
```go
type td struct {
        TotalDifficulty *hexutil.Big `json:"totalDifficulty"`
}
var td *TotalDifficulty
err := ec.c.CallContext(ctx, &td, "eth_getBlockByNumber", "latest", false)
if err == nil && td == nil {
	panic("not found")
}
```

Reply:
Added in latest commit, thanks!